core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationTimerTask.java

```diff
@@ -0,0 +1,101 @@
+ */
+package org.sonarsource.sonarlint.core.notifications;
+
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.TimerTask;
+import java.util.function.BinaryOperator;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonarsource.sonarlint.core.client.api.common.NotificationConfiguration;
+import org.sonarsource.sonarlint.core.client.api.connected.ServerConfiguration;
+import org.sonarsource.sonarlint.core.client.api.notifications.SonarQubeNotification;
+
+class NotificationTimerTask extends TimerTask {
+  // merge with most recent time
+  private static final BinaryOperator<ZonedDateTime> MERGE_TIMES = (t1, t2) -> t1.toInstant().compareTo(t2.toInstant()) > 0 ? t1 : t2;
+  private static final Logger LOG = LoggerFactory.getLogger(NotificationTimerTask.class);
+  private final NotificationCheckerFactory checkerFactory;
+  private Collection<NotificationConfiguration> configuredProjects = Collections.emptyList();
+
+  public NotificationTimerTask() {
+    this(new NotificationCheckerFactory());
+  }
+
+  public NotificationTimerTask(NotificationCheckerFactory checkerFactory) {
+    this.checkerFactory = checkerFactory;
+  }
+
+  public void setProjects(Collection<NotificationConfiguration> configurations) {
+    this.configuredProjects = new ArrayList<>(configurations);
+  }
+
+  @Override
+  public void run() {
+    Map<ServerConfiguration, List<NotificationConfiguration>> mapByServer = groupByServer();
+
+    for (Map.Entry<ServerConfiguration, List<NotificationConfiguration>> entry : mapByServer.entrySet()) {
+      requestForServer(entry.getKey(), entry.getValue());
+    }
+  }
+
+  private static ZonedDateTime getLastNotificationTime(NotificationConfiguration config) {
+    ZonedDateTime lastTime = config.persistence().getLastNotificationTime();
+    ZonedDateTime oneDayAgo = ZonedDateTime.now().minusDays(1);
+    return lastTime.isAfter(oneDayAgo) ? lastTime : oneDayAgo;
+  }
+
+  private void requestForServer(ServerConfiguration serverConfiguration, List<NotificationConfiguration> configs) {
+    try {
+      Map<String, ZonedDateTime> request = configs.stream()
+        .collect(Collectors.toMap(NotificationConfiguration::projectKey, NotificationTimerTask::getLastNotificationTime, MERGE_TIMES));
+
+      NotificationChecker notificationChecker = checkerFactory.create(serverConfiguration);
+      List<SonarQubeNotification> notifications = notificationChecker.request(request);
+
+      for (SonarQubeNotification n : notifications) {
+        Stream<NotificationConfiguration> matchingConfStream = configs.stream();
+        if (n.projectKey() != null) {
+          matchingConfStream = matchingConfStream.filter(c -> c.projectKey().equals(n.projectKey()));
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor the code so this stream pipeline is used. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3958)